libraries/chain/block_state.cpp

```diff
@@ -86,7 +88,69 @@ std::pair<bool, std::optional<uint32_t>> block_state::aggregate_vote(const vote_
       return {};
    }
 }
-   
+
+// Called from net threads
+void block_state::verify_qc(const valid_quorum_certificate& qc) const {
+   const auto& finalizers = active_finalizer_policy->finalizers;
+   auto num_finalizers = finalizers.size();
+
+   // utility to accumulate voted weights
+   auto weights = [&] ( const hs_bitset& votes_bitset ) -> uint64_t {
+      uint64_t sum = 0;
+      auto n = std::min(num_finalizers, votes_bitset.size());
+      for (auto i = 0u; i < n; ++i) {
+         if( votes_bitset[i] ) { // ith finalizer voted
+            sum += finalizers[i].weight;
+         }
+      }
+      return sum;
+   };
+
+   // compute strong and weak accumulated weights
+   auto strong_weights = qc._strong_votes ? weights( *qc._strong_votes ) : 0;
+   auto weak_weights = qc._weak_votes ? weights( *qc._weak_votes ) : 0;
+
+   // verfify quorum is met
+   if( qc.is_strong() ) {
+      EOS_ASSERT( strong_weights >= active_finalizer_policy->threshold,
+                  block_validate_exception,
+                  "strong quorum is not met, strong_weights: ${s}, threshold: ${t}",
+                  ("s", strong_weights)("t", active_finalizer_policy->threshold) );
+   } else {
+      EOS_ASSERT( strong_weights + weak_weights >= active_finalizer_policy->threshold,
+                  block_validate_exception,
+                  "weak quorum is not met, strong_weights: ${s}, weak_weights: ${w}, threshold: ${t}",
+                  ("s", strong_weights)("w", weak_weights)("t", active_finalizer_policy->threshold) );
+   }
+
+   std::vector<bls_public_key> pubkeys;
+   std::vector<std::vector<uint8_t>> digests;
+
+   // utility to aggregate public keys and digests for verification
+   auto prepare_pubkeys_digests = [&] ( const auto& votes_bitset, const auto& digest ) {
+      auto n = std::min(num_finalizers, votes_bitset.size());
+      pubkeys.reserve(n);
+      digests.reserve(n);
+      for (auto i = 0u; i < n; ++i) {
+         if( votes_bitset[i] ) { // ith finalizer voted the digest
+            pubkeys.emplace_back(finalizers[i].public_key);
+            digests.emplace_back(std::vector<uint8_t>{digest.data(), digest.data() + digest.data_size()});
+         }
+      }
+   };
+
+   // aggregate public keys and digests for strong and weak votes
+   if( qc._strong_votes ) {
+      prepare_pubkeys_digests(*qc._strong_votes, strong_digest);
+   }
+   if( qc._weak_votes ) {
+      prepare_pubkeys_digests(*qc._weak_votes, weak_digest);
+   }
```

Comment:
```cpp
if (qc._strong_votes) {
   pubkeys.emplace_back(aggregate_pubkeys(*qc._strong_votes));
   digests.emplace_back(std::vector<uint8_t>{strong_digest.data(), strong_digest.data() + strong_digest.data_size()});
}

if (qc._weak_votes) {
   pubkeys.emplace_back(aggregate_pubkeys(*qc._weak_votes));
   digests.emplace_back(std::vector<uint8_t>{weak_digest.data(), weak_digest.data() + weak_digest.data_size()});
}
```

This way, by the time you execute `aggregate_verify`, `pubkeys` and `digests` will both only either have size 1 or size 2.

Reply:
Great! Thanks!